hubble/src/indexer/fetcher.rs

```diff
@@ -0,0 +1,212 @@
+                        catch_up_range.start, catch_up_range.end
+                    );
+
+                    for slice in catch_up_range.range_chunks(self.chunk_size).into_iter() {
+                        info!("handling chunk: {}..{}", slice.start, slice.end);
+
+                        let (tx, mut rx) = mpsc::channel(self.chunk_size);
+                        last_finalized
+                            .fetch_range(tx, slice.clone())
+                            .instrument(info_span!("fetch"))
+                            .await?;
+
+                        let mut expected_next_height = slice.start;
+
+                        while let Some(message) = rx.recv().await {
+                            expected_next_height = self
+                                .handle_fetch_message(message, expected_next_height)
+                                .instrument(info_span!("store"))
+                                .await?
+                        }
+                        info!("{}..{}: handled chunk", &slice.start, &slice.end);
+                    }
+                }
+            };
+        }
+    }
+
+    async fn handle_fetch_message(
+        &self,
+        message: FetchMessage<T::BlockHandle>,
+        expected_next_height: BlockHeight,
+    ) -> Result<BlockHeight, Report> {
+        match message {
+            FetchMessage::Fetched(block_handle) => {
+                debug!("fetched: {}", block_handle.height()?);
+
+                if expected_next_height != block_handle.height()? {
+                    error!(
+                        "unexpected: {} (expecting {})",
+                        block_handle.height()?,
+                        expected_next_height
+                    );
+                    return Err(FetcherError::RecievedUnexpectedHeight(
+                        block_handle.height()?,
+                        expected_next_height,
+                    )
+                    .into());
+                }
+
+                let mut tx = self.pg_pool.begin().await?;
+
+                block_handle.insert(&mut tx).await?;
+
+                update_current_height(
+                    &mut tx,
+                    self.indexer_id.clone(),
+                    block_handle.height()?,
+                    block_handle.timestamp()?,
+                )
+                .await?;
+
+                tx.commit().await?;
+
+                Ok(expected_next_height + 1)
+            }
+            FetchMessage::Error(error) => {
+                error!("error: {}", error);
+                Err(FetcherError::ErrorFetchingMessage(error).into())
+            }
+        }
+    }
+
+    async fn run_to_tip(&self, fetcher_client: &T) -> Result<(), Report> {
+        loop {
+            let next_height = self.next_height().await?;
+            info!("{}: fetching", next_height);
+
+            match fetcher_client
+                .fetch(BlockReference::Height(next_height))
+                .await?
+            {
```

Comment:
So nit, this now does it sequentially again (we're no longer fetching in batches and inserting as fast as they come in). That actually means syncing from scratch takes a very long time (note some chains produces blocks sub 1 second, so the amount of blocks is quite big and keeps growing very quickly).

Reply:
we still have this approach:
- first: run to finalized
- then: run to tip